lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+	 * If we're looking for "any tag" then this property will be `null`.
+	 *
+	 * h1...h6 are special since they are variations of the same base tag.
+	 * To find "any heading tag" pass the special value `h`.
+	 *
+	 * @var string|null
+	 */
+	public $tag_name;
+
+
+	/**
+	 * We're looking for a tag also containing this CSS class name, up to
+	 * the comparable equivalence of those names. If we're not looking for
+	 * a class name this property will be `null`.
+	 *
+	 * @var string|null
+	 */
+	public $class_name;
+
+
+	/**
+	 * We're looking for a tag that also has some data-attribute values.
+	 * If all we care about is the presence of one of those values this
+	 * will contain the name of that attribute, up to the comparable
+	 * equivalence of that name. If we also want to constrain the value
+	 * of that attribute we will have a pair containing the name of the
+	 * attribute (to equivalence) with a predicate function. If we're
+	 * not looking for a data-attribute this property will be `null`.
+	 *
+	 * Example:
+	 * <code>
+	 *     // has a `data-wp-container` attribute
+	 *     'data-wp-container'
+	 *
+	 *     // has a `data-wp-container` attribute with the value `core/group`
+	 *     ['data-wp-block', 'core/group']
+	 *
+	 *     // has a `data-test-group` attribute with an even-valued id
+	 *     ['data-test-group', function ( $group ) { return 0 === (int) $group % 2; }]
+	 * </code>
+	 *
+	 * @var array<string|callable>
+	 */
+	public $data_attribute;
+
+
+	/**
+	 * Used to skip matches in case we expect more than one to exist.
+	 * This constraint applies after all other constraints have held.
+	 * For example, to find the third `<div>` tag containing the
+	 * `wp-block` class name set `match_offset = 3`. If not provided
+	 * the default indication is to find the first match.
+	 *
+	 * @default 1
+	 * @var int
+	 */
+	public $match_offset;
+
+
+	/**
+	 * Creates a tag find descriptor given the input parameters specifying
+	 * the intended match, encodes inputs for searching.
+	 *
+	 * @param array $query {
+	 *     Which tag name to find, having which class, etc…
+	 *
+	 *     @type string|null $tag_name   Which tag to find, or `null` for "any tag," or `h` for "any h1 through h6."
+	 *     @type int|null $match_offset  Find the Nth tag matching all search criteria.
+	 *                                   1 for "first" tag, 3 for "third," etc.
+	 *                                   Defaults to first tag.
+	 *     @type string|null $class_name Tag must contain this whole class name to match.
+	 *     @type array<string|callable>  Tag must contain data-attribute of given name and optionally a given
+	 *                                   value, or a given predicate function which returns whether the
+	 *                                   attribute's value constitutes a match.
+	 * }
+	 * @return WP_Tag_Find_Descriptor Used by WP_HTML_Processor when scanning HTML.
+	 */
+	public static function parse( $query ) {
```

Comment:
Any reason this can't be a constructor instead of a static method?

Reply:
probably, but at the time I wasn't sure if we'd want to create them through other means. I think there are still some questions I have about the role here. for instance, I do think if this were a construction I'd want to still call it like so

```php
public function __construct( $query = array() ) {
    $parsed = self::parse( $query );
    $this->tag_name = $parsed->tag_name;
    …
}
```

and at that point I don't see what the value is in having the constructor. my original intent was to make the `parse` function failable, and maybe there's still a point in doing that, but for now it fails to safe defaults silently.

if this is the only way we'd want to create a descriptor then maybe we just convert it to a constructor, but I think for some purposes we might imagine creating one manually

```php
$d = new Descriptor();
$d->tag_name = '<invalid';
```